### DISCUSSION TITLE
ModuleFederation helpers in separate library

### DISCUSSION THREAD

**Discussion by NEWESTERS** [webpack/webpack#15962] [Q&A]
I have some helper functions like this from webpack's examples:
```js
function loadComponent(scope, module) {
  return async () => {
    await __webpack_init_sharing__("default");
    const container = window[scope];
    await container.init(__webpack_share_scopes__.default);
    const factory = await window[scope].get(module);
    const Module = factory();
    return Module;
  };
}
```
Copy-pasting this snippet between all federated applications is not good idea and i decided to extract them to separate library, that may contain also non-javascript assets, so i use __webpack__ to build this library.

But if i understand correctly, __webpack__ replaces internal variables like `__webpack_init_sharing__` during build time, so they doesn't exist in target application and helpers doesn't work.

Is there workarounds for this behavior or should i build helpers library with another bundler (e.g. __rollup__), that doesn't know about webpack's internal variables?

**Comment by vankop** [comment]
replace list for such replacements https://github.com/webpack/webpack/blob/main/lib/RuntimeGlobals.js

**Reply by NEWESTERS** [reply]
Thanks for answer!

Should i do something like this?
```js
const { RuntimeGlobals } = require("webpack");

RuntimeGlobals.initializeSharing = "__webpack_init_sharing__";
```

Is this approach safe?

**Reply by vankop** [reply]
not, you should not redefine them, you should use right value

**Reply by NEWESTERS** [reply]
Should i use `__webpack_require__.I` instead of `__webpack_init_sharing__` in my code?
AFASIK webpack replaces `__webpack_init_sharing__` with `__webpack_require__.I` when i build library and then library code doesn't work in application.
But if i build library with rollup, it works fine.

**Reply by vankop** [reply]
please create small repro.. i don't get it

**Reply by NEWESTERS** [reply]
I've tried to reproduce this behavior in minimal example, but it works as it should, seems like problem has gone.
Maybe i've done something wrong in my previous attempt.

Here's implementation of my idea: <https://github.com/NEWESTERS/module-federation-example>.
There is package named `library` with its own webpack config.
Source code of `library` uses webpack's internal variable `__webpack_init_sharing__`, which will be replaced by webpack on `library` build script.
Then i run webpack in `host` application, that depends on `library`, and that means that `library` source code went through webpack twice.
## TL;DR
So abstract version of my question is: "What happens, if source code, that contains webpack's  internal variables, goes through webpack twice (__source code__ -> __webpack__ -> __build-1__ -> __webpack__ -> __build-2__)".

**Reply by vankop** [reply]
if you minimize result, e.g. by default in production, comments will be removed
